plugins/producer_plugin/include/eosio/producer_plugin/block_timing_util.hpp

```diff
@@ -0,0 +1,49 @@
+#pragma once
+#include <eosio/chain/block_timestamp.hpp>
+#include <eosio/chain/config.hpp>
+
+namespace eosio {
+
+enum class pending_block_mode { producing, speculating };
+
+struct block_timing_util {
+   int32_t cpu_effort_us = 0;
+   fc::time_point pending_block_deadline;
```

Comment:
caching `pending_block_deadline` is required with the current design because the value is dependent on the when it is calculated (i.e. `time_point::now()`) while `process_incoming_transaction_async()` and `push_read_only_transaction()` require it. If we don't cache it and recalcuate it like before,  the deadlines used by those two member function will be different from those used in `start_block()`.